pylint/extensions/bad_chained_comparison.py

```diff
@@ -0,0 +1,47 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from astroid import nodes
+
+from pylint.checkers import BaseChecker
+
+if TYPE_CHECKING:
+    from pylint.lint import PyLinter
+
+COMPARISON_OP = frozenset(("<", "<=", ">", ">=", "!=", "=="))
+IDENTITY_OP = frozenset(("is", "is not"))
+MEMBERSHIP_OP = frozenset(("in", "not in"))
+
+
+class BadChainedComparisonChecker(BaseChecker):
+    """Checks for unintentional usage of chained comparison."""
+
+    name = "bad-chained-comparison"
+    msgs = {
+        "W2401": (
+            "Expression gets interpreted as a %s-part chained comparison which straddles comparison groups. If this is not the intent, please parenthesize.",
+            "bad-chained-comparison",
+            "Used when there is a chained comparison where one expression is part of two comparisons that belong to different groups.",
+        )
+    }
+
+    def _has_diff_comparison_groups(self, node: nodes.Compare) -> bool:
+        operators = [op[0] for op in node.ops]
+        # Check if comparison operators are in the same group
+        for comparison_group in [COMPARISON_OP, IDENTITY_OP, MEMBERSHIP_OP]:
+            if operators[0] in comparison_group:
+                group = comparison_group
+        for operator in operators:
+            if operator not in group:
+                return True
+        return False

```

Comment:
```suggestion
        return any(o not in group for o in operators)
```
Look like a false negative of the for-any-all checker.

Reply:
I think both options should have the same performance?

Let's say `o not in group for o in operators` evaluates to the following iterable: `[False, True, True]`.
Then, `o in group for o in operators` evaluates to the inverse: `[True, False, False]`.

`any([False, True, True])` short circuits at the first True, and conversely, `all([True, False, False])`  short circuits at the first False.

My point is, I don't think the frequency of operators that are not in the group matters, since the short circuit effect should always occur on the same index in both iterables.

Please correct me if I am missing something here.

Reply:
Refactored according to pylint's suggestions.

Reply:
You're right it's identical (Actually I initially made the same mistake in https://github.com/PyCQA/pylint/issues/4146). Resolving, sorry about the noise 😄